fractal_tasks_core/lib_ngff.py

```diff
@@ -0,0 +1,412 @@
+            ratio_y = round(
+                self.pixel_sizes_zyx[ind][1] / self.pixel_sizes_zyx[ind - 1][1]
+            )
+            if ratio_x != ratio_y:
+                raise NotImplementedError(
+                    "Inhomogeneous coarsening in X/Y directions "
+                    "is not supported.\n"
+                    f"ZYX pixel sizes:\n {self.pixel_sizes_zyx}"
+                )
+            if current_ratio is None:
+                current_ratio = ratio_x
+            else:
+                if current_ratio != ratio_x:
+                    raise NotImplementedError(
+                        "Inhomogeneous coarsening across levels "
+                        "is not supported.\n"
+                        f"ZYX pixel sizes:\n {self.pixel_sizes_zyx}"
+                    )
+
+        return current_ratio
+
+
+class Image(BaseModel):
+    """
+    Model for an element of `Well.images`.
+
+    See https://ngff.openmicroscopy.org/0.4/#well-md.
+
+    NOTE: we deviate from NGFF specs, since we allow `path` to be an arbitrary
+    string.
+    TODO: include a check like `constr(regex=r'^[A-Za-z0-9]+$')`, through a
+    Pydantic validator.
+    """
+
+    acquisition: Optional[int] = Field(
+        None, description="A unique identifier within the context of the plate"
+    )
+    path: str = Field(
+        ..., description="The path for this field of view subgroup"
+    )
+
+
+class Well(BaseModel):
+    """
+    Model for `NgffWellMeta.well`.
+
+    See https://ngff.openmicroscopy.org/0.4/#well-md.
+    """
+
+    images: list[Image] = Field(
+        ...,
+        description="The images included in this well",
+        min_items=1,
+        unique_items=True,
+    )
+    version: Optional[str] = Field(
+        None, description="The version of the specification"
+    )
+
+
+class NgffWellMeta(BaseModel):
+    """
+    Model for the metadata of a NGFF well.
+
+    See https://ngff.openmicroscopy.org/0.4/#well-md.
+    """
+
+    well: Optional[Well] = None
+
+    def get_acquisition_paths(self) -> dict[int, str]:
+        """
+        Create mapping from acquisition indices to corresponding paths.
+
+        Runs on the well zarr attributes and loads the relative paths in the
+        well.
+
+        Returns:
+            Dictionary with `(acquisition index: image path)` key/value pairs.
+
+        Raises:
+            ValueError:
+                If an element of `self.well.images` has no `acquisition`
+                    attribute.
+            NotImplementedError:
+                If acquisitions are not unique.
+        """
+        acquisition_dict = {}
+        for image in self.well.images:
+            if image.acquisition is None:
+                raise ValueError(
+                    "Cannot get acquisition paths for Zarr files without "
+                    "'acquisition' metadata at the well level"
+                )
+            if image.acquisition in acquisition_dict:
+                raise NotImplementedError(
```

Comment:
Is this a Fractal or a spec constraint? Let's make this more clear.

Reply:
Fixed with d6d21db9dc38e13be8d3658b70ba474da3bf1798.